README.md
# python-abx #

<http://github.com/lrvick/python-abx>

## About ##

A simple GTK application for performing ABX testing using python and gstreamer.

This codebase is a slightly cleaned up fork of [python-abx](https://code.google.com/archive/p/python-abx/) which was abandoned, and itself was a fork of the long
dead [ABX-Comparator](http://freecode.com/projects/abx-comparator)

## Requirements ##
  * python2.7
  * GTK2
  * gstreamer-0.10 (And various good, bad, and ugly modules as needed)
  * python-gtk
  * python-gst

## Current Features ##

  * Swap A/B on at any point while playing
  * Perform any number of tests and show/hide results at any time

## Usage / Installation ##

1. run abx.py

    ```bash
    ./abx.py
    ```
2. Select files for A and B

3. Play A,B,X and make make your guesses

4. "Show" to show results

## Notes ##

  I only tweaked this enough to get it working for me.
  Use at your own risk. You may be eaten by a grue.


ChangeLog
2012-05-26  Adam Fontenot <adam.m.fontenot@gmail.com>
    
    * Complete rewrite
    * Contained entirely in abx.py

2006-06-19  Артём Попов  <artfwo@gmail.com>

	* Released version 0.2

2006-06-19  Артём Попов  <artfwo@gmail.com>

	* Looper.py: seeking is now fixed
	* Looper.py: will not play video from now
	* Cosmetic changes here and there

2006-04-14  Артём Попов  <artfwo@gmail.com>

	* Initial release


abx2.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-

import sys
import gtk
import gst
import random
import gobject
import math

# the gtk class
class AbxComparator:

    # basic variables for the program
    # sound_player gives us a playbin2 instance
    sound_player = gst.element_factory_make("playbin2", "sound_player")

    a_location, a_duration, b_duration, b_location = None, None, None, None
    current_location = None

    correct, incorrect = 0, 0

    begin_sample, end_sample, current_sample = 0, 0, 0

    # keep track of whether we're dragging the position bar
    mouse_active_on_hscale = False

    # set the locations if available
    def __init__(self, a_location=None, b_location=None):

        ui = "abx-comparator.ui"
        builder = gtk.Builder()
        builder.add_from_file(ui)
        builder.connect_signals(self)
        window = builder.get_object("abx_audio_window")
        window.show_all()

        # define elements of the .ui so we can access them
        self.a_button = builder.get_object("a_button")
        self.b_button = builder.get_object("b_button")
        self.x_button = builder.get_object("x_button")
        self.isa_button = builder.get_object("isa_button")
        self.isb_button = builder.get_object("isb_button")
        self.stop_button = builder.get_object("stop_button")
        self.repeat_button = builder.get_object("repeat_button")
        self.audio_position = builder.get_object("audio_position")
        self.audio_adjustment = builder.get_object("adjustment1")
        self.show_results_button = builder.get_object("show_results_button")
        self.text_buffer = builder.get_object("resultsview").get_buffer()
        self.begin_button = builder.get_object("startbutton")
        self.end_button = builder.get_object("endbutton")

    # a few obvious handlers
    def _on_main_window_delete_event(self, *args):
        self.quit()

    def _on_quit_button_clicked(self, *args):
        self.quit()

    def _on_show_results_button_toggled(self, *args):
        self.update_results()

    def _on_clear_results_button_clicked(self, *args):
        self.correct, self.incorrect = 0, 0
        self.update_results()

    # the _on_*_button_toggled handlers handle playback and pausing.
    # TODO may want to move some of these functions later
    def _on_a_button_toggled(self, *args):
        if self.a_button.get_active():
            # if paused, unpause
            if self.sound_player.get_state()[1] == gst.STATE_PAUSED:
                self.sound_player.set_state(gst.STATE_PLAYING)
                gobject.timeout_add(50, self.update_slider)
                return
            # if we're playing now on a different button, set a place holder for the position so we can switch it immediately
            if self.sound_player.get_state()[1] == gst.STATE_PLAYING:
                self.current_sample = self.sound_player.query_position(gst.FORMAT_TIME)[0]
            else:
                self.current_sample = self.begin_sample
            self.phoenix()
            self.x_button.set_active(False)
            self.b_button.set_active(False)
            self.sound_player.set_property('uri', self.a_location)
            self.sound_player.set_state(gst.STATE_PLAYING)
            # workaround for a playbin2 bug - we have to get the current state before it starts playing for some reason. check upstream?
            self.sound_player.get_state()
            self.sound_player.seek_simple(gst.FORMAT_TIME, gst.SEEK_FLAG_FLUSH, self.current_sample)
            self.begin_button.set_sensitive(True)
            self.end_button.set_sensitive(True)
            # set a 50 msec timeout to update the position slider
            gobject.timeout_add(50, self.update_slider)
        else:
            self.sound_player.set_state(gst.STATE_PAUSED)

    # the other buttons are similar...
    def _on_b_button_toggled(self, *args):
        if self.b_button.get_active():
            if self.sound_player.get_state()[1] == gst.STATE_PAUSED:
                self.sound_player.set_state(gst.STATE_PLAYING)
                gobject.timeout_add(50, self.update_slider)
                return
            if self.sound_player.get_state()[1] == gst.STATE_PLAYING:
                self.current_sample = self.sound_player.query_position(gst.FORMAT_TIME)[0]
            else:
                self.current_sample = self.begin_sample
            self.phoenix()
            self.a_button.set_active(False)
            self.x_button.set_active(False)
            self.sound_player.set_property('uri', self.b_location)
            self.sound_player.set_state(gst.STATE_PLAYING)
            self.sound_player.get_state()
            self.sound_player.seek_simple(gst.FORMAT_TIME, gst.SEEK_FLAG_FLUSH, self.current_sample)
            self.begin_button.set_sensitive(True)
            self.end_button.set_sensitive(True)
            gobject.timeout_add(50, self.update_slider)
        else:
            self.sound_player.set_state(gst.STATE_PAUSED)

    def _on_x_button_toggled(self, *args):
        if self.x_button.get_active():
            if self.sound_player.get_state()[1] == gst.STATE_PAUSED:
                self.sound_player.set_state(gst.STATE_PLAYING)
                gobject.timeout_add(50, self.update_slider)
                return
            if self.sound_player.get_state()[1] == gst.STATE_PLAYING:
                self.current_sample = self.sound_player.query_position(gst.FORMAT_TIME)[0]
            else:
                self.current_sample = self.begin_sample
            self.phoenix()
            self.a_button.set_active(False)
            self.b_button.set_active(False)
            self.sound_player.set_property('uri', self.current_location)
            self.sound_player.set_state(gst.STATE_PLAYING)
            self.sound_player.get_state()
            self.sound_player.seek_simple(gst.FORMAT_TIME, gst.SEEK_FLAG_FLUSH, self.current_sample)
            self.begin_button.set_sensitive(True)
            self.end_button.set_sensitive(True)
            gobject.timeout_add(50, self.update_slider)
        else:
            self.sound_player.set_state(gst.STATE_PAUSED)
    
    # if user IDs a song, stop playing, and check for correct, giving points accordingly
    def _on_isa_button_clicked(self, *args):
        self.stop()
        if self.current_location == self.a_location: 
            self.correct += 1
        else:
            self.incorrect += 1
        self.update_gui()

    def _on_isb_button_clicked(self, *args):
        self.stop()
        if self.current_location == self.b_location: 
            self.correct += 1
        else:
            self.incorrect += 1
        self.update_gui()

    # set positions to loop over section of audio
    def _start_selection_button_toggled(self, *args):
        if self.begin_button.get_active():
            # set rounded version of time as label text
            self.begin_button.set_label(str("%.2f" %(self.sound_player.query_position(gst.FORMAT_TIME)[0] / gst.MSECOND / 1000.0)))
            # record position to begin_sample
            self.begin_sample = self.sound_player.query_position(gst.FORMAT_TIME)[0]
        else:
            self.begin_button.set_label("Start")
            self.begin_sample = 0

    def _end_selection_button_toggled(self, *args):
        if self.end_button.get_active():
            self.end_button.set_label(str("%.2f" %(self.sound_player.query_position(gst.FORMAT_TIME)[0] / gst.MSECOND / 1000.0)))
            self.end_sample = self.sound_player.query_position(gst.FORMAT_TIME)[0]
            # we've just set an "end position", so immediately loop to the begin_sample
            self.sound_player.seek_simple(gst.FORMAT_TIME, gst.SEEK_FLAG_FLUSH, self.begin_sample)
        else:
            self.end_button.set_label("End")
            self.end_sample = 0

    def _on_stop_button_clicked(self, *args):
        self.stop()

    def _a_file_chosen(self, button):
        self.a_location = button.get_uri()
        self.a_duration = self.load_file(self.a_location)
        self.update_gui()

    def _b_file_chosen(self, button):
        self.b_location = button.get_uri()
        self.b_duration = self.load_file(self.b_location)
        self.update_gui()

    # set the bool for mouse dragging when we're active
    def _hscale_button_press(self, *args):
        self.mouse_active_on_hscale = True

    def _hscale_button_release(self, *args):
        # seek to the new location
        self.sound_player.seek_simple(gst.FORMAT_TIME, gst.SEEK_FLAG_FLUSH, self.audio_adjustment.value)
        # and then set the bool. not the other way around to avoid potential problems with timing
        self.mouse_active_on_hscale = False
        return
        
    def load_file(self, location):
        # load a temporary playbin2 instance, and use it to determine duration
        tempPlayer = gst.element_factory_make("playbin2", "tempPlayer")
        tempPlayer.set_property('uri', location)
        tempPlayer.set_state(gst.STATE_PLAYING)
        tempPlayer.get_state()

        try:
	        duration = tempPlayer.query_duration(gst.FORMAT_TIME)[0]
        except:
	        print("Couldn't determine stream duration.\n",
                                    "Unsupported stream?")	

        tempPlayer.set_state(gst.STATE_NULL)

        print location, duration
        return duration

    def update_gui(self, *args):
        # only run if both files are loaded
        if (self.a_duration != None and self.b_duration != None):
            # set pos. slider to min of two files to prevent user from "cheating"
            self.audio_adjustment.upper = min(self.a_duration, self.b_duration)
            self.enable_buttons()
            # prepare test
            if random.random() > 0.5:
                self.current_location = self.a_location
            else:
                self.current_location = self.b_location
        self.update_results()

    def update_results(self, *args):
        if self.show_results_button.get_active():
            score_value = str(self.correct)
            
            # hackish way of calculating p
            # p = 0.5 ^ (number of trials) * summation of the binominal coefficient from (number of guesses) to (number of attempts)
            # see http://mathforum.org/library/drmath/view/71267.html and http://vassarstats.net/textbook/index.html chapter 5 append. 5

            n = self.correct + self.incorrect
            k = self.correct
            p = 0

            for x in range(k, n + 1):
                p = p + math.factorial(n) / (math.factorial(x) * math.factorial(n - x)) * 0.5 ** n
            p_str = "%.2f%%" % round(p * 100, 2)
        else:
            score_value, p_str = "hidden", "hidden"
        
        # set the oddly named text_buffer of the GtkTextBox so the user can see his score
        self.text_buffer.set_text("Score: " + score_value + " / " + str(self.correct + self.incorrect) + "\np = " + str(p_str))

    # run on a timer every 50msec whenever sound_player is playing
    def update_slider(self, *args):
        # don't deactivate if dragging position slider, just bypass
        if self.mouse_active_on_hscale:
            return True
        # if we're not playing, stop the timer
        if self.sound_player.get_state()[1] == gst.STATE_NULL:
            return False
        # if we've run past the slider or past the user's limit, stop
        if self.sound_player.query_position(gst.FORMAT_TIME)[0] > self.audio_adjustment.upper or self.sound_player.query_position(gst.FORMAT_TIME)[0] > self.end_sample and self.end_sample != 0:
            #repeat if selected
            if self.repeat_button.get_active():
                # sanity check the begin_sample, and go there if appropriate
                if self.end_sample > self.begin_sample:
                    self.sound_player.seek_simple(gst.FORMAT_TIME, gst.SEEK_FLAG_FLUSH, self.begin_sample)
                else:
                    self.sound_player.seek_simple(gst.FORMAT_TIME, gst.SEEK_FLAG_FLUSH, 0.0)
            else:
                self.stop()
                return False
        # seems to fail occasionally if still seeking, so simply get_state() and wait for the next timed update
        # TODO: is get_state() actually doing anything?
        try:
            self.audio_adjustment.value = self.sound_player.query_position(gst.FORMAT_TIME)[0]
        except:
            self.sound_player.get_state()
        return self.sound_player.get_state()[1] == gst.STATE_PLAYING

    # phoenix(), a dirty hack to work around bugs in playbin2
    # idea from the quodlibet player
    # see http://code.google.com/p/quodlibet/source/browse/quodlibet/quodlibet/player/gstbe.py

    def phoenix (self, *args):
        self.sound_player.set_state(gst.STATE_NULL)
        self.sound_player.get_state()
        self.sound_player = None
        self.sound_player = gst.element_factory_make("playbin2", "sound_player")
     
    # and a few obvious handlers   
    def enable_buttons(self, *args):
        self.a_button.set_sensitive(True)
        self.x_button.set_sensitive(True)
        self.b_button.set_sensitive(True)
        self.isa_button.set_sensitive(True)
        self.isb_button.set_sensitive(True)
        self.stop_button.set_sensitive(True)
        self.audio_position.set_sensitive(True)

    def stop(self, *args):
        self.a_button.set_active(False)
        self.b_button.set_active(False)
        self.x_button.set_active(False)
        self.begin_button.set_sensitive(False)
        self.end_button.set_sensitive(False)
        self.sound_player.set_state(gst.STATE_NULL)
        self.audio_adjustment.value = 0
    
    def quit(self, *args):
        gtk.main_quit(*args)

# "MAIN" CODE STARTS HERE

# set window icon
gtk.window_set_default_icon_from_file("abx-comparator.svg")

a, b = None, None
# if we have at least 2 arguments, go ahead and set file A
if 2 < len(sys.argv) <= 3:
    a = sys.argv[1]
    if not gst.uri_is_valid(a):
        a = "file://" + os.path.abspath(a)

# and if we have 3, set file B as well
if len(sys.argv) == 3:
    b = sys.argv[2]
    if not gst.uri_is_valid(b):
        b = "file://" + os.path.abspath(b)

# define the app class
app = AbxComparator(a, b)
# launch the window
gtk.main()


abxer/Metadata.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
#
# ABX Comparator
# Copyright (c) 2006 Артём Попов <artfwo@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#

import pygst
pygst.require("0.10")
import gst

import pygtk
pygtk.require("2.0")
import gobject

from gst.extend import discoverer

class Metadata:
    def __init__(self, location):
        self.__metadata = {}
        player = gst.element_factory_make("playbin2", "player")
        player.set_property('uri', location)
        player.set_state(gst.STATE_PLAYING)
        player.get_state()
	try:
	    self.__metadata["duration"] = player.query_duration(gst.FORMAT_TIME)[0]
	except:
	    self.__error = ("Couldn't determine stream duration.",
                                    "Unsupported stream?")	
	player.set_state(gst.STATE_NULL)

    def __getitem__(self, key):
        return self.__metadata[key]


abxer/__init__.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
#
# ABX Comparator
# Copyright (c) 2006 Артём Попов <artfwo@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#

from Metadata           import *
from Looper             import *
from SampleSelector     import *


abxer/SampleSelector.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
#
# ABX Comparator
# Copyright (c) 2006 Артём Попов <artfwo@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#

import pygtk
pygtk.require("2.0")
import gtk

(SAMPLESEL_DRAG_NONE,
SAMPLESEL_DRAG_LEFT,
SAMPLESEL_DRAG_RIGHT,
SAMPLESEL_DRAG_ALL) = range(4)

SAMPLESEL_DRAG_ZONE = 2

class SampleSelector(gtk.DrawingArea):
    def __init__(self):
        gtk.DrawingArea.__init__(self)

        self.add_events(gtk.gdk.BUTTON_PRESS_MASK |
                gtk.gdk.BUTTON_RELEASE_MASK |
                gtk.gdk.POINTER_MOTION_MASK)

        self.connect("button-press-event", self.__on_button_press_event)
        self.connect("button-release-event", self.__on_button_release_event)
        self.connect("motion-notify-event", self.__on_motion_notify_event)
        self.connect("expose_event", self.__on_expose_event)

        self.set_size_request(9, 9) # maybe even bigger?

        self.mode = SAMPLESEL_DRAG_NONE

        self.start_value        = 0.0
        self.stop_value         = 1.0
        self.playback_position  = 0.0

    def start_pixels(self):
        return round(self.get_allocation().width * self.start_value)

    def stop_pixels(self):
        return round(self.get_allocation().width * self.stop_value)

    def __in_left_handle(self, position):
        return self.start_pixels() - SAMPLESEL_DRAG_ZONE < position < \
            self.start_pixels() + SAMPLESEL_DRAG_ZONE

    def __in_right_handle(self, position):
        return self.stop_pixels() - SAMPLESEL_DRAG_ZONE < position < \
            self.stop_pixels() + SAMPLESEL_DRAG_ZONE

    def __in_between(self, position):
        return self.start_pixels() + SAMPLESEL_DRAG_ZONE < position < \
            self.stop_pixels() - SAMPLESEL_DRAG_ZONE

    def __on_button_press_event(self, widget, event):
        if event.button != 1:
            return

        if self.__in_left_handle(event.x):
            self.mode = SAMPLESEL_DRAG_LEFT
        elif self.__in_right_handle(event.x):
            self.mode = SAMPLESEL_DRAG_RIGHT
        elif self.__in_between(event.x):
            self.window.set_cursor(gtk.gdk.Cursor(gtk.gdk.FLEUR))
            self.mode = SAMPLESEL_DRAG_ALL
        else:
            return

        self.x1 = event.x

    def __on_button_release_event(self, widget, event):
        if event.button != 1:
            return

        # a little update if we have moved anything
        if (self.mode != SAMPLESEL_DRAG_NONE):
            self.window.set_cursor(None)
            del self.x1
            self.mode = SAMPLESEL_DRAG_NONE

    def __on_motion_notify_event(self, widget, event):
        if event.is_hint:
            x, y, state = event.window.get_pointer()
        else:
            x = event.x
            y = event.y
            state = event.state

        # default behaviour, change cursor
        if self.mode == SAMPLESEL_DRAG_NONE:
            if self.__in_left_handle(x):
                self.window.set_cursor(gtk.gdk.Cursor(gtk.gdk.LEFT_SIDE))
            elif self.__in_right_handle(x):
                self.window.set_cursor(gtk.gdk.Cursor(gtk.gdk.RIGHT_SIDE))
            else:
                self.window.set_cursor(None)
            return

        dx = x - self.x1
        dv = float(dx) / self.get_allocation().width

        # left handle
        if self.mode == SAMPLESEL_DRAG_LEFT:
            new_start = self.start_value + dv
            if state & gtk.gdk.SHIFT_MASK:
                new_stop = self.stop_value - dv
            else:
                new_stop = self.stop_value

            if new_start < 0: new_start = 0
            if new_stop > 1: new_stop = 1

            if new_start < new_stop:
                self.start_value = new_start
                self.stop_value = new_stop
                self.queue_draw()
            self.x1 = event.x

        # right handle
        elif self.mode == SAMPLESEL_DRAG_RIGHT:
            new_stop = self.stop_value + dv
            if state & gtk.gdk.SHIFT_MASK:
                new_start = self.start_value - dv
            else:
                new_start = self.start_value

            if new_start < 0: new_start = 0
            if new_stop > 1: new_stop = 1

            if new_stop > new_start:
                self.start_value = new_start
                self.stop_value = new_stop
                self.queue_draw()
            self.x1 = event.x

        # blue thingie
        elif self.mode == SAMPLESEL_DRAG_ALL:
            new_start = self.start_value + dv
            new_stop = self.stop_value + dv

            # for smooth collision within limits
            if new_start < 0:
                new_stop -= new_start
                new_start = 0
            elif new_stop > 1:
                new_start -= new_stop - 1
                new_stop = 1

            self.start_value = new_start
            self.stop_value = new_stop
            self.queue_draw()

            self.x1 = event.x

    def __on_expose_event(self, widget, event):
        style = self.rc_get_style()

        if self.get_property("sensitive"):
            bg_color = style.base[gtk.STATE_NORMAL]
            fg_color = style.fg[gtk.STATE_NORMAL]
            selection_color = style.bg[gtk.STATE_SELECTED]
            tracker_color = style.base[gtk.STATE_NORMAL]
        else:
            bg_color = style.bg[gtk.STATE_INSENSITIVE]
            # fg_color = style.fg[gtk.STATE_INSENSITIVE]
            fg_color = style.fg[gtk.STATE_NORMAL]
            # selection_color = style.text[gtk.STATE_INSENSITIVE]
            selection_color = style.bg[gtk.STATE_SELECTED]
            tracker_color = style.base[gtk.STATE_NORMAL]

        context = widget.window.cairo_create()
        context.rectangle(event.area.x,
            event.area.y,
            event.area.width,
            event.area.height)
        context.clip()

        w = self.get_allocation().width
        h = self.get_allocation().height

        self.__draw_background(context, bg_color, fg_color)
        self.__draw_range(context, selection_color, fg_color)

        if self.playback_position > 0:
            self.__draw_position_indicator(context, tracker_color)

        return False

    def __draw_background(self, context, bg_color, fg_color):
        context.rectangle(0.5, 0.5,
                self.get_allocation().width - 1,
                self.get_allocation().height - 1)

        context.set_line_width(1.0)

        context.set_source_rgb(bg_color.red / float(0xffff),
                bg_color.green / float(0xffff),
                bg_color.blue / float(0xffff))

        context.fill_preserve()

        context.set_source_rgb(fg_color.red / float(0xffff),
                fg_color.green / float(0xffff),
                fg_color.blue / float(0xffff))

        context.stroke()

        context.rectangle(0.5, 2.5,
                self.get_allocation().width - 1,
                self.get_allocation().height - 5)

        context.set_source_rgb(fg_color.red / float(0xffff),
                fg_color.green / float(0xffff),
                fg_color.blue / float(0xffff))

        context.stroke()

    def __draw_range(self, context, bg_color, fg_color):
        context.rectangle(self.start_pixels() + 0.5,
                2.5,
                self.stop_pixels() - self.start_pixels() - 1,
                self.get_allocation().height - 5)

        context.set_line_width(1.0)

        context.set_source_rgb(bg_color.red / float(0xffff),
                bg_color.green / float(0xffff),
                bg_color.blue / float(0xffff))

        context.fill_preserve()

        context.set_source_rgb(fg_color.red / float(0xffff),
                fg_color.green / float(0xffff),
                fg_color.blue / float(0xffff))

        context.stroke()

    def __draw_position_indicator(self, context, fg_color):
        x = self.get_allocation().width * self.playback_position

        context.set_line_width(1.0)
        context.move_to(x, 2.5)
        context.line_to(x, self.get_allocation().height - 2.5)

        context.set_source_rgb(fg_color.red / float(0xffff),
                fg_color.green / float(0xffff),
                fg_color.blue / float(0xffff))

        context.stroke()


abxer/Looper.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
#
# ABX Comparator
# Copyright (c) 2006 Артём Попов <artfwo@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#

import pygst
pygst.require("0.10")
import gst

import pygtk
pygtk.require("2.0")
import gobject

class Looper(gobject.GObject):
    __gproperties__ = {
        "loop": (gobject.TYPE_BOOLEAN,
            "loop",
            "Whether to loop the segment",
            False,
            gobject.PARAM_READWRITE),
        "start-pos": (gobject.TYPE_UINT64,
            "start position",
            "The segment start marker",
            0,
            0xfffffffffffffff, # max long possible
            0,
            gobject.PARAM_READWRITE),
        "stop-pos": (gobject.TYPE_UINT64,
            "stop position",
            "The segment stop marker",
            0,
            0xfffffffffffffff, # max long possible
            0,
            gobject.PARAM_READWRITE),
    } # __gproperties__

    __gsignals__ = {
        "stopped": (gobject.SIGNAL_RUN_LAST,
            gobject.TYPE_NONE, ()),
        "position-updated": (gobject.SIGNAL_RUN_LAST,
            gobject.TYPE_NONE, (gobject.TYPE_FLOAT,)),
        "error": (gobject.SIGNAL_RUN_LAST,
            gobject.TYPE_NONE, (gobject.TYPE_PYOBJECT,))
    } # __gsignals__

    def __init__(self, location=None):
        gobject.GObject.__init__(self)

        self.__playbin = gst.element_factory_make("playbin")
        self.__playbin.props.video_sink = gst.element_factory_make("fakesink")

        bus = self.__playbin.get_bus()
        bus.add_watch(self.__on_bus_message)

        self.__loop = False
        self.__start_pos    = 0
        self.__stop_pos     = 0
        self.__timeout_id   = 0

        if location:
            self.load(location)

    def load(self, location):
        self.__playbin.props.uri = location
        self.__start_position = 0
        self.__stop_position = 0

    def set_segment(self, start, stop):
        self.props.start_pos = start
        self.props.stop_pos = stop

    def play(self):
        if not (self.__start_pos or self.__stop_pos):
            raise RuntimeError, "Cannot start playback, segment was not set!"

        self.__playbin.set_state(gst.STATE_PLAYING)

    def stop(self, silent=False):
        self.__playbin.set_state(gst.STATE_NULL)
        if not silent:
            self.emit("stopped")

    def is_playing(self):
        state_change, state, pending = self.__playbin.get_state()
        # FIXME: need to implement a better check here
        return state == gst.STATE_PLAYING or state == gst.STATE_PAUSED

    def do_get_property(self, property):
        if property.name == "loop":
            return self.__loop
        elif property.name == "start-pos":
            return self.__start_pos
        elif property.name == "stop-pos":
            return self.__stop_pos
        else:
            raise AttributeError, "Unknown property %s" % property.name

    def do_set_property(self, property, value):
        if property.name == "loop":
            self.__loop = value
        elif property.name == "start-pos":
            self.__start_pos = value
        elif property.name == "stop-pos":
            self.__stop_pos = value
        else:
            raise AttributeError, "Unknown property %s" % property.name

    def do_stopped(self):
        if self.__timeout_id:
            gobject.source_remove(self.__timeout_id)
            self.__timeout_id = 0

    def __seek(self, start, stop, flush):
        flags = gst.SEEK_FLAG_SEGMENT | gst.SEEK_FLAG_ACCURATE
        if flush:
            flags = flags | gst.SEEK_FLAG_FLUSH
        self.__playbin.seek(1.0, gst.FORMAT_TIME, flags,
                gst.SEEK_TYPE_SET, start,
                gst.SEEK_TYPE_SET, stop)

    def __on_timeout(self):
        try:
            position = self.__playbin.query_position(gst.FORMAT_TIME)[0]
            self.emit("position-updated", float(position))
        except:
            return False
        return True

    def __on_bus_message(self, bus, message):
        if message.type == gst.MESSAGE_ERROR:
            error, debug = message.parse_error()
            self.stop() # this seems neccessary here
            self.emit("error", (error, debug))

        elif message.type == gst.MESSAGE_NEW_CLOCK:
            # we connect the timeout handler here
            # to be sure that further queries succeed
            interval = int((self.__stop_position - self.__start_position) / \
                (2 * gst.SECOND) + 50)
            self.__timeout_id = gobject.timeout_add(interval,
                self.__on_timeout)

        elif message.type == gst.MESSAGE_STATE_CHANGED:
            old_state, new_state, pending = message.parse_state_changed()
            if (old_state == gst.STATE_READY and
                new_state == gst.STATE_PAUSED and
                message.src == self.__playbin):
                self.__seek(self.__start_pos, self.__stop_pos, True)

        elif message.type == gst.MESSAGE_SEGMENT_DONE:
            if self.__loop:
                self.__seek(self.__start_pos, self.__stop_pos, False)
            else:
                src = self.__playbin.get_property("source")
                pad = src.get_pad('src')
                pad.push_event(gst.event_new_eos())

                # this is the good old way:
                #
                # pads = src.src_pads()
                # while True:
                #       try:
                #               pad = pads.next()
                #               pad.push_event(gst.event_new_eos())
                #       except:
                #               break

        elif message.type == gst.MESSAGE_EOS:
            self.stop()

        return True
